pkg/controllers/teamrbac/teamrolebinding_controller.go

```diff
@@ -191,38 +175,133 @@ func (r *TeamRoleBindingReconciler) Reconcile(ctx context.Context, req ctrl.Requ
-				return ctrl.Result{}, err
+	trbStatus.SetConditions(greenhousev1alpha1.TrueCondition(greenhousev1alpha1.RBACReady, greenhousev1alpha1.RBACReconciled, ""))
+
+	return ctrl.Result{}, nil
+}
+
+// cleanupResources removes rbacv1 resources from all clusters that are no longer matching the the TeamRoleBinding's clusterSelector/clusterName
+func (r *TeamRoleBindingReconciler) cleanupResources(ctx context.Context, trbStatus greenhousev1alpha1.TeamRoleBindingStatus, trb *greenhousev1alpha1.TeamRoleBinding, clusters *greenhousev1alpha1.ClusterList) (greenhousev1alpha1.TeamRoleBindingStatus, error) {
+	for _, s := range trbStatus.PropagationStatus {
+		// remove rbac for all clusters no longer matching the clusterSelector
+		if !slices.ContainsFunc(clusters.Items, func(c greenhousev1alpha1.Cluster) bool { return c.GetName() == s.ClusterName }) {
+			cluster := &greenhousev1alpha1.Cluster{}
+			err := r.Get(ctx, types.NamespacedName{Namespace: trb.Namespace, Name: s.ClusterName}, cluster)
+			if apierrors.IsNotFound(err) {
+				// cluster has been removed, nothing to be done
+				trbStatus = removePropagationStatus(trbStatus, s.ClusterName)
+				continue
 			}
+			if err != nil {
+				return trbStatus, err
+			}
+			if err = r.cleanupCluster(ctx, trb, cluster); err != nil {
+				return trbStatus, err
+			}
+			trbStatus = removePropagationStatus(trbStatus, s.ClusterName)
 		}
 	}
+	return trbStatus, nil
+}
 
-	// TODO: add status update logic here
-	return ctrl.Result{}, nil
+// cleanupCluster removes the TeamRoleBinding's rbacv1 resources from the cluster, returns an error if the cleanup fails.
+// This will remove the rbacv1.ClusterRoleBinding if the TeamRoleBinding is not namespaced, otherwise it will remove all rbacv1.RoleBindings
+// If the rbacv1.ClusterRole is no longer referenced, it will be removed as well
+func (r *TeamRoleBindingReconciler) cleanupCluster(ctx context.Context, trb *greenhousev1alpha1.TeamRoleBinding, cluster *greenhousev1alpha1.Cluster) error {
+	cl, err := getK8sClient(ctx, r.Client, cluster)
+	if err != nil {
+		log.FromContext(ctx).Error(err, "Error getting client for cluster %s to delete TeamRoleBinding %s", cluster.GetName(), trb.GetName())
+		return err
+	}
+
+	if len(trb.Spec.Namespaces) > 0 {
+		if err := r.deleteRoleBindings(ctx, cl, trb, cluster); err != nil {
+			return err
+		}
+	} else {
+		if err := r.deleteClusterRoleBinding(ctx, cl, trb, cluster); err != nil {
+			return err
+		}
+	}
+
+	// if all RBAC bindings for this TeamRoleBinding have been deleted, check if the Role is still needed
+	// if the rbacv1.ClusterRole is not referenced by any other rbacv1.ClusterRole, rbacv1.Role, it can be removed from the remote cluster
+	isReferenced, err := isRoleReferenced(ctx, cl, trb)
+	switch {
+	case err != nil:
+		return err
+	case isReferenced:
+		return nil
+	}
+
+	// ClusterRole is no longer referenced, delete the ClusterRole
```

Comment:
The CRDs provide no complete list of clusters that a TeamRole is deployed to. 
This requires either

- looking at `.status.propagationStatus` of all TeamRoleBindings (might not be up to date)
- using TeamRoleBinding's `.spec.clusterName` to list all clusters and then check if the cluster is contained by any

On the remote cluster this can be achieved with the list request utilising the dedicated labels. Also this makes sure that we do not have to rely on the `propagationStatus` but can check with the actual resources if the `ClusterRole` can be removed.